javalib/src/main/scala/java/util/StringJoiner.scala

```diff
@@ -0,0 +1,76 @@
+// Ported from Scala.js commit: 57d71da dated: 2023-05-31
+// Two methods are different, see below.
+
+package java.util
+
+@inline
+final class StringJoiner private (
+    delimiter: String,
+    prefix: String,
+    suffix: String
+) extends AnyRef {
+
+  /** The custom value to return if empty, set by `setEmptyValue` (nullable).
+   *
+   *  If `null`, defaults to `prefix + suffix`.
+   */
+  private var emptyValue: String = null
+
+  /** The current value, excluding prefix and suffix. */
+  private var value: String = ""
```

Comment:
It's works nice in JS, but for native/jvm we can do better. The main operation for adding new string is `add(CharSequence)`. Upon String materialization in `toString` we need to add custom prefix and suffix. Instead of constructing intermiediete string we can store parts of it comming from `add`. We can use the fact the List prepend in Scala is very cheap. Here's how I think it could be implemented to more efficient in jvm/native. We have extreamly fast add, and slower materializaiton

```scala
private def delimiter, prefix, suffix: CharSequence 
private def emptyValue: CharSequence | Null 
private var parts: List[CharSequence] = Nil 

def add(v: CharSequence) = { parts ::= v; this } 
def toString() = {
 if parts.isEmpty && emptyValue != null 
 then emptyValue 
 else {
   val sb = new java.lang.StringBuilder(this.length()) // alloc exactly as much buffer as required to skip internal buffer resize
   sb.append(prefix) 
   val head :: tail = parts.reverse(): @unchecked // we know parts is non-empty, parts are in reverse order
   sb.append(head)
   tail.foreach{ part => sb.append(delimiter).append(part) }
   sb.append(suffix)
   sb.toString()
 }
 def merge(other: StringJoiner): StringJoiner = { 
    // take all the parts from other and join 
    if (!other.isEmpty){
      parts = other.parts ::: this.parts
    }
    this
 }
 def length(): Int = {
   val baseLength = if(emptyValue != null) emptyValue.length else prefix.length + suffix.length() 
   val delimLength = delimiter.length() 
   parts.foldLeft(baseLength){case (acc, part) => acc + delimLength + part.length }
 }
}
```

Reply:
Is `parts.reverse()` (followed by `.foreach`) costly here? Could we use `parts.foldRight` maybe? Is that better, or about the same?

Reply:
It looks like, with `merge`, we have to use the delimiter of `other` for parts that come from `other`:
```scala
scala> val multiple = StringJoiner(";", "[", "]").setEmptyValue("--").add("a").add("b").add("c")
val multiple: java.util.StringJoiner = [a;b;c]
                                                                                                                      
scala> val joiner = StringJoiner(", ", "{", "}").add("one").merge(multiple).add("two")
val joiner: java.util.StringJoiner = {one, a;b;c, two}
// here a;b;c is using delimiter of multiple, not of joiner
```
So the `parts = other.parts ::: this.parts` approach doesn't work (or maybe I'm not seeing something obvious). 
What should I do?

Reply:
Oh, ok, that's oversight from me. So docs state that the other.parts are concatanate using other delimieter, and it should not add prefix/suffix or empty value. So we can do either do 1 one of things here: 
 1. flatMap other.parts to create a list of delimiters and parts and then prepend them to this.parts 
 2. create new StringBuilder (which is also an instance of CharSequence) and materialize the other joiner without prefix/suffix or empty value. For that we can just copy-past code from toString with minor modifications, or even refector them out to common function. I think this solution might be better.

Reply:
I think I kinda understand what you're saying. 
I'll have to figure out the string length of (`other.parts` + the delimiters that go between them) for the new `StringBuilder`, I might also refactor that into another helper function and reuse it in `length()`.